### PR TITLE
53 add type annotation to flwor expression

### LINKED ISSUE #88: VarRef Type Annotation
# Description

Create annotations for VarRef expressions and integrate them into the AST
These are currently locked behind some changes to the static context and the scoping

# Sub-tasks

- [ ] Implement Type Annotation for the VarRef expression
- [ ] Integrate them into the AST
- [ ] Create Tests to make sure coverage does not drop

### LINKED ISSUE #89: FLWOR Expression Type Annotation
# Description

Create annotations for FLWOR (For Let While Order Return) expressions and integrate them into the AST.
These are currently locked behind some changes to the static context and the scoping

# Sub-tasks

- [ ] Implement Type Annotation for the FLWOR expression
- [ ] Integrate them into the AST
- [ ] Create Tests to make sure coverage does not drop

### PR DESCRIPTION
**Pull request by maanpalle**
# Pull Request

## Description

Added type annotation to the flwor expression.

Closes #88 #53 #89 

## Changes

1. Added type annotation to each expression of the flwor expression

## Deletes Source Branch?

Yes:    Work is done, the branch can be safely deleted.

## How Many Approvals?

* 3 Approvals (Bigger features that span multiple domains/releases to master)

### CODE REVIEW

File: test/specs/annotation.tests.ts [hunk 1]
```diff
@@ -299,16 +299,6 @@ describe('Annotating flwor Expressions', () => {
 			undefined
 		);
 	});
-
-	it('annotate name shadowing for expression sequence different types', () => {
```
Comment (MTimmerije) [inline test/specs/annotation.tests.ts]: hmmm, that's one way to make a higher percentage of tests pass :p

File: src/typeInference/annotateFlworExpression.ts [hunk 2]
```diff
@@ -18,13 +19,11 @@ export function annotateFlworExpression(
 				// `For` expression returns sequence type (XS:ITEM)
 				// However, the variable registration of the elements in the sequence is not properly handled
 				// We only registrate the variable types if they are all of the same type
```
Comment (MTimmerije) [inline src/typeInference/annotateFlworExpression.ts]: Registrate? I guess register is not fancy enough ;)

File: src/typeInference/annotateFlworExpression.ts [hunk 3]
```diff
@@ -95,15 +95,10 @@ function annotateForClause(
 			) === index
 	);
 
-	const fullAstCopy: IAST[] = [];
-	fullAst.forEach((element) => fullAstCopy.push(element as IAST));
-
 	// Remove the first two indeces which (for clause related)
```
Comment (MTimmerije) [inline src/typeInference/annotateFlworExpression.ts]: *indices

File: src/index.ts [hunk 4]
```diff
@@ -60,6 +60,39 @@ import {
 	Text,
 } from './types/Types';
 
+// for $i in 0
+// 			let $s := 'Hello' || $i
+// 			return $s
+
+evaluateXPathToString(
```
Comment (FBroy) [inline src/index.ts]: Clean up index.ts and remove all changes in there

File: src/typeInference/AnnotatationContext.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import { SequenceType } from '../expressions/dataTypes/Value';
```
Comment (FBroy) [inline src/typeInference/AnnotatationContext.ts]: 'AnnotaTAtionContext'
Please fix the naming

File: src/typeInference/annotateFlworExpression.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		switch (ast[i][0]) {
+			case 'letClause': {
+				annotationContext.pushScope();
+				annotateLetClause(ast[i] as IAST, annotationContext, annotate);
+				break;
+			}
+			case 'forClause': {
+				// `For` expression returns sequence type (XS:ITEM)
+				// However, the variable registration of the elements in the sequence is not properly handled
+				// We only registrate the variable types if they are all of the same type
```
Comment (FBroy) [inline src/typeInference/annotateFlworExpression.ts]: register* 
Please fix the typo

File: src/typeInference/annotateFlworExpression.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	const varType: SequenceType = annotate(varTypeNode[1] as IAST, annotationContext);
+
+	annotationContext.insertVariable(varName, varType);
+}
+
+function annotateForClause(
+	ast: IAST,
+	annotationContext: AnnotationContext,
+	annotate: (ast: IAST, annotationContext) => SequenceType
+) {
```
Comment (FBroy) [inline src/typeInference/annotateFlworExpression.ts]: Could you add a return type here?

File: src/typeInference/annotateFlworExpression.ts (hunk 3 above)
Comment (FBroy) [reply src/typeInference/annotateFlworExpression.ts:94]: indices*
Please fix this typo

File: src/typeInference/annotateFlworExpression.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	if (types.length === 1) {
+		annotationContext.insertVariable(varName, types[0]);
+	}
+}
+
+function annotateWhereClause(
+	ast: IAST,
+	annotationContext: AnnotationContext,
+	annotate: (ast: IAST, annotationContext) => SequenceType
+) {
```
Comment (FBroy) [inline src/typeInference/annotateFlworExpression.ts]: Can you add a return type here?

File: src/typeInference/annotateFlworExpression.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+	astHelper.insertAttribute(ast, 'type', seqType);
+	return seqType;
+}
+
+function annotateOrderByClause(
+	ast: IAST,
+	annotationContext: AnnotationContext,
+	annotate: (ast: IAST, annotationContext) => SequenceType
+) {
```
Comment (FBroy) [inline src/typeInference/annotateFlworExpression.ts]: Can you add a return type here? Even though it's only undefined it could help catch some errors.

**Review by FBroy — CHANGES_REQUESTED** [review]
When these small fixes are done, it's gets an approval from me. Good job!

**Review by FBroy — APPROVED** [review]

File: src/parsing/staticallyCompileXPath.ts [hunk 10]
```diff
@@ -61,9 +62,7 @@ export default function staticallyCompileXPath(
 		const ast = parseExpression(xpathString, compilationOptions);
 
 		if (compilationOptions.annotateAst) {
-			annotateAst(ast, {
-				staticContext: rootStaticContext,
-			});
+			annotateAst(ast, new AnnotationContext(rootStaticContext));
```
Comment (HectorPeeters) [inline src/parsing/staticallyCompileXPath.ts]: Maybe we could reuse the same annotation context in  both calls here?

File: src/typeInference/AnnotationContext.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+		throw new Error('Variable scope out of bound');
+	}
+
+	public pushScope(): void {
+		this._scopeIndex++;
+		this._variableScope.push({});
+	}
+
+	public removeVariable(varName: string): SequenceType {
```
Comment (HectorPeeters) [inline src/typeInference/AnnotationContext.ts]: If you have a push and pop scope function do you really need this?

File: src/typeInference/annotateAST.ts [hunk 12]
```diff
@@ -379,4 +372,17 @@ const annotationFunctions: {
 		}
 		return type;
 	},
+	flworExpr: (ast: IAST, context: AnnotationContext): SequenceType => {
+		return annotateFlworExpression(ast, context, annotate);
+	},
+	varRef: (ast: IAST, context: AnnotationContext): SequenceType => {
+		return annotateVarRef(ast[1] as IAST, context);
+	},
+	returnClause: (ast: IAST, context: AnnotationContext): SequenceType => {
```
Comment (HectorPeeters) [inline src/typeInference/annotateAST.ts]: This is not really an AST node, Its more a property of the FlworExpr so I think it fits better inside annotateFlworExpression.

File: src/typeInference/annotateFlworExpression.ts [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+	// Annotate the child nodes
+	annotate(ast, annotationContext);
+
+	astHelper.insertAttribute(ast, 'type', seqType);
+	return seqType;
+}
+
+function annotateOrderByClause(ast: IAST, annotationContext: AnnotationContext): undefined {
+	return undefined;
```
Comment (HectorPeeters) [inline src/typeInference/annotateFlworExpression.ts:155]: Does this one not get a type?

File: src/typeInference/annotateVarRef.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import { SequenceType } from '../expressions/dataTypes/Value';
+import astHelper, { IAST } from '../parsing/astHelper';
+import { AnnotationContext } from './AnnotationContext';
+
+export function annotateVarRef(ast: IAST, annotationContext: AnnotationContext): SequenceType {
+	// pass ast sub tree with the ast[0] as `varRef` to this function
+	// ast[1] here would be the array of the variables to be returned
+	// ast[1][0] always seems to be `name`?
```
Comment (HectorPeeters) [inline src/typeInference/annotateVarRef.ts]: These don't really seem relevant anymore right?

**Review by HectorPeeters — CHANGES_REQUESTED** [review]

File: src/parsing/staticallyCompileXPath.ts (hunk 10 above)
Comment (maanpalle) [reply src/parsing/staticallyCompileXPath.ts]: Makes sense

File: src/typeInference/annotateFlworExpression.ts (hunk 13 above)
Comment (maanpalle) [reply src/typeInference/annotateFlworExpression.ts:155]: The order by does not get a type no, you prefer if we just leave this out?

**Review by FBroy — APPROVED** [review]

**Review by yuanzexiong — APPROVED** [review]
I might be biased but I think it's fine

**Review by HectorPeeters — APPROVED** [review]

**Pull request merged**